compiler/rustc_typeck/src/check/type_ascription.rs

```diff
@@ -0,0 +1,142 @@
+use super::FnCtxt;
+use crate::expr_use_visitor::{ConsumeMode, Delegate, ExprUseVisitor};
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::def::Res;
+use rustc_middle::hir::place::PlaceWithHirId;
+use rustc_middle::ty;
+use rustc_span::hygiene::SyntaxContext;
+use rustc_span::symbol::Ident;
+use rustc_span::Span;
+
+// Checks for type ascriptions in lvalue contexts, i.e. in situations such as
+// (x : T) = ..., &(x : T), &mut (x : T) and (x : T).foo(...), where
+// type ascriptions can potentially lead to type unsoundness problems
+struct TypeAscriptionValidator<'a, 'tcx> {
+    fcx: &'a FnCtxt<'a, 'tcx>,
+}
+
+impl<'a, 'tcx> Delegate<'tcx> for TypeAscriptionValidator<'a, 'tcx> {
+    fn consume(
+        &mut self,
+        _place_with_id: &PlaceWithHirId<'tcx>,
+        _diag_expr_id: hir::HirId,
+        _mode: ConsumeMode,
+    ) {}
+
+    fn borrow(
+        &mut self,
+        place_with_id: &PlaceWithHirId<'tcx>,
+        _diag_expr_id: hir::HirId,
+        borrow_expr: Option<&hir::Expr<'_>>,
+        bk: ty::BorrowKind,
+    ) {
+        debug!(
+            "TypeAscriptionkind::borrow(place: {:?}, borrow_expr: {:?}, borrow_kind: {:?})",
+            place_with_id.place, borrow_expr, bk
+        );
+
+        if let Some(borrow) = borrow_expr {
```

Comment:
This is the right general idea, but I'm worried it's not quite covering all the bases. This will work for something like

```rust
&(x:T)
```

but I don't think it would work for `&(x:T).f`. I'll try to create what seem to me to be a good set of test cases and link them below. 

I'm also not sure if looking at `borrow_expr` is the right thing; the `Place` has a `HirId`, maybe we want to look at that. I have to check under what conditions `borrow_expr` is None.

Reply:
Update: OK, I see you added `borrow_expr`. I think instead of doing that, you can use [`expect_expr`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.expect_expr) to get the `hir::Expr`. 

To manage the cases I was talking about, you'll need to use a HIR visitor or something to recursively walk the HIR structure.

I am, however, pondering whether the right fix here is to modify the `Place` structure a bit. Let me take a look at that.

Reply:
So yes, what I think is that we should modify place construction -- I suspect we should 

* modify [`PlaceBase::Rvalue`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/place/enum.PlaceBase.html#variant.Rvalue) to contain a `HirId`
* treat ascription as an rvalue, during place construction but...probably just everywhere -- that means modifying [this function](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Expr.html#method.is_place_expr) and perhaps [this one](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Expr.html#method.is_syntactic_place_expr), I haven't read that much into that code
* check in this code where `place.base` is an rvalue and, if so, check whether its an ascription and report an error

Reply:
See comment above for why I introduced `borrow_expr`. I also tried to somehow (currently don't remember what exactly I tried) to get access to a parent expression (the type ascription expression) for `diag_expr_id`, but wasn't able to do that. 

I will take a look at your suggestions later.

Reply:
I took a look at `mem_categorization` and it isn't easily evident to me why we would get a different expression, maybe I need to spend some more time. The reason the interface includes a `diagnostics_hir_id` is so that we can support providing a hir id that is different from HirId in `PlaceWithHirId` which I think can be utilized here along with `expect_expr`. Assuming that there is a valid reason for `x: T` to not contain the later part of that expression.  (It's currently being only used for destructure patterns).